Automattic/woocommerce-payments
# WIP: Introduce a class to schedule upgrade subscription data migrations

Fixes #2894
Fixes #2880

### Changes proposed in this Pull Request

This PR adds a base class that is responsible for:

- Recognising when the WC Subscriptions plugin has been activated or deactivated. To enable that, this class houses various functions for:
     - Checking if the store is currently in the process of upgrading.
     - Validating the upgrade status. Upgrade -> the plugin is still active.
     - Kicking off the background batch processors when an upgrade needs to occur.
     - Initiating the background processors so their hooks are listening to the Action Scheduler events. 
- Adds a background batch processing class to upgrade WCPay Subscriptions via Action Scheduler

### Testing instructions

⚠️ **Following these steps will result in an irreversible data migration.** ⚠️ 
**After 1 hour, all your Stripe Billing Subscriptions will be cancelled.**

1. Make sure you have multiple WCPay/Stripe Billing Subscriptions active on your site.
3. Activate the **WC Subscriptions** plugin. 
4. Go to **Tools > Action Scheduler** and search for `'wc_payments_schedule_wcpay_subscription_upgrades'`. 
5. You should see at least 1 action scheduled and depending on how fast you are or how busy the queue is, it will either be **pending** or **completed**. (https://d.pr/i/ZFmiE4)
     - When this action runs, it will find subscriptions that need to be upgraded and schedule individual actions.
6. Once that action has been completed. Search the action scheduler queue for `'wc_payments_tokenize_wcpay_subscription'`. https://d.pr/i/FyGfpS
     - Here you will find all the individual upgrade actions for each WCPay subscription. 
     - Note they are scheduled to run in ~1 hour.
6. Make a note of one of the subscriptions in the queue and find the corresponding subscription in the Stripe dashboard. 
    - Note that it is active. https://d.pr/i/lXKWM4
7. View the specific subscription in the WC edit subscription screen. 
   - Note the billing schedule section is locked (uneditable) https://d.pr/i/gHLdHm
   - Note the line items cannot be edited https://d.pr/i/9NE9jA
7. Click **run** on one of those actions. https://d.pr/i/AVwFB3
8. Refresh the subscription in the Stripe dashboard and notice it is now cancelled. https://d.pr/i/x6Ct8B
8. Refresh the edit subscription screen in WC. Note everything is editable. https://d.pr/i/Go3F65, https://d.pr/i/ETiGTG
9. Search your database options table for the `'wc_pay_subscription_migration_status'` key and notice it is set to `'upgrading'` https://d.pr/i/NOff3e 
9. Deactivate the **WC Subscriptions** plugin. 
10. Search your database options table for the `'wc_pay_subscription_migration_status'` key now - there should be 0 results.
10. Search the scheduled action page for the migration actions again. They should now all be cancelled. https://d.pr/i/A1b7P0

---

**Reactivating the WC Subscriptions extension**

1. Activate the **WC Subscriptions** plugin. 
2. The process defined above will restart. 
    - `'wc_payments_schedule_wcpay_subscription_upgrades'` actions leading to `wc_payments_tokenize_wcpay_subscription` actions.
4. Search your database options table for the `'wc_pay_subscription_migration_status'` key and notice it is set to `'upgrading'` https://d.pr/i/NOff3e 
5. This time we're going to let it run so we'll need to wait an hour. 
6. Once all the actions have been processed, if you look in your Stripe dashboard, all your subscriptions should be cancelled https://d.pr/Z37aKg
7. Checking the database for the `'wc_pay_subscription_migration_status'` key option again, and it will be deleted too. The upgrade is finished! 

-------------------

**Testing renewals of upgraded subscriptions**

1. Go to one of your upgraded subscription's edit admin screens. 
2. Attempt to process a renewal from the actions dropdown. (https://d.pr/i/48Bs7N)
3. Payment should be taken as if it was a tokenised WC Payments subscription. 
   - I confirmed this by just seeing the transaction in the Stripe dashboard. 
  
----


**Post merge**
